_posts/2019-11-15-enforcing-database-transactions-with-rust.md

```diff
@@ -0,0 +1,214 @@
+`thing-id` is deleted, the output of `db/fetch-thing` can get quite
+unpredictable. Could we make this kind of programs safer?
+
+## Enter Rust
+
+Rust is a relatively new programming language, with focus on speed and safety.
+Rust has the unique notion of 'ownership', which enables memory safety without
+relying on a garbage collector. Nowadays in most languages all variables are
+actually references, which are just thrown around and are either wildly mutable
+or completely immutable. Rust provides a way to write code where mutation is
+explicit and safe.
+
+Rust implements this by providing an unique feature: the borrow checker.
+Essentially, the borrow checker is a compile-time feature which ensures that
+objects live long enough, and at the same time prevents unsafe concurrent access
+to variables. This is implemented by following a couple simple rules:
+
+1. If an object drops out of scope, it is destroyed
+2. An object must outlive all references to it
+3. An object can only have either multiple immutable references, or a single mutable reference
+
+In this blog post we are mostly concerned about the first rule -- if an object
+drops out of scope, it is destroyed.
+
+The borrow checker is mostly praised for its ability to enable manual memory
+management without suffering from null pointer exceptions or segfaults -- Safe
+Rust doesn't have any null or otherwise invalid references! In my opinion, the
+best benefit of the borrow checker is not the speedup it provides by
+eliminating the garbage collector, but the ability to catch logic bugs. If you
+manage to construct your problem as a problem around ownership, the Rust
+compiler can check your logic at compile-time!
+
+## Trait interlude
+
+Rust doesn't have inheritance nor interfaces. Instead, generic code is
+implemented around *traits*. Traits are similar to interfaces in the sense that
+they contain a bunch of methods, and certain types implement those interfaces.
+Types can only implement traits either where the type is defined, or where the
+trait is defined. This unfortunately means that one cannot implement a
+third-party trait for a third-party type. However, unlike interfaces in eg.
+Java, one can implement their own traits for other types.
+
+When writing generic code, generic type parameters can receive trait bounds to
+specify things that can be done with the types. For example, in the following
+function
+
+```rust
+fn debug_print<T: Debug>(t: T) { dbg!(t); }
+```
+
+the type T is required to implement the trait `Debug`, which allows turning the
+object into an programmer-readable (but not necessarily human-readable) format.
+
+The generic functions are compiled similarily as in C++: any calls to the
+function are specialized to the types, which technically can create binary
+bloat, but practically reduces it as the compiler has better options for
+optimization.
+
+## Modeling database queries around ownership
+
+So, we need to be able to
```

Comment:
Stylistic issue, I would expand the three things as text and not have a numbered list here... ymmv.